packages/block-editor/src/store/actions.js

```diff
@@ -1432,36 +1439,128 @@ export const setNavigationMode =
+
+			if ( mode === 'zoom-out' || prevMode === 'zoom-out' ) {
+				const sectionsContainerClientId = unlock(
+					registry.select( STORE_NAME )
+				).getSectionsContainerClientId();
+
+				if ( sectionsContainerClientId ) {
+					// If we have a sections container (usually a core/group
+					// <main> element), only allow editing of the contents of
+					// the container.
+					const sectionsClientIds = select.getClientIdsOfDescendants(
+						sectionsContainerClientId
+					);
+					if ( mode === 'zoom-out' ) {
+						dispatch.setBlockEditingMode(
+							ROOT_CONTAINER_CLIENT_ID,
+							'disabled'
+						);
+						dispatch.setBlockEditingMode(
+							sectionsContainerClientId,
+							'contentOnly'
+						);
+						sectionsClientIds.forEach( ( clientId ) =>
+							dispatch.setBlockEditingMode( clientId, 'default' )
+						);
+					} else if ( prevMode === 'zoom-out' ) {
+						dispatch.unsetBlockEditingMode(
+							ROOT_CONTAINER_CLIENT_ID
+						);
+						dispatch.unsetBlockEditingMode(
+							sectionsContainerClientId
+						);
+						sectionsClientIds.forEach( ( clientId ) => {
+							dispatch.unsetBlockEditingMode( clientId );
+						} );
+					}
+				} else {
+					// If we don't have a sections container, we get the top-level
+					// blocks and only allow editing those blocks.
+					const sectionsClientIds = select.getBlockOrder();
+					const disabledSectionsClientIds = sectionsClientIds.filter(
+						( clientId ) => {
+							const block = select.getBlock( clientId );
+							return block.name === 'core/template-part';
+						}
+					);
+					const sectionClientIdsInnerBlocks =
+						sectionsClientIds.flatMap( ( clientId ) => {
+							const block = select.getBlock( clientId );
+							if ( block.name === 'core/template-part' ) {
+								return [];
+							}
+							return block.innerBlocks.map(
+								( innerBlock ) => innerBlock.clientId
+							);
+						} );
+					if ( mode === 'zoom-out' ) {
+						dispatch.setBlockEditingMode(
+							ROOT_CONTAINER_CLIENT_ID,
+							'contentOnly'
+						);
+						sectionsClientIds.forEach( ( clientId ) =>
+							dispatch.setBlockEditingMode(
+								clientId,
+								'contentOnly'
+							)
+						);
+						disabledSectionsClientIds.forEach( ( clientId ) =>
+							dispatch.setBlockEditingMode( clientId, 'disabled' )
+						);
+						sectionClientIdsInnerBlocks.forEach( ( clientId ) =>
+							dispatch.setBlockEditingMode( clientId, 'disabled' )
+						);
+					} else if ( prevMode === 'zoom-out' ) {
+						dispatch.unsetBlockEditingMode(
+							ROOT_CONTAINER_CLIENT_ID
+						);
+						sectionsClientIds.forEach( ( clientId ) =>
+							dispatch.unsetBlockEditingMode( clientId )
+						);
+						disabledSectionsClientIds.forEach( ( clientId ) =>
+							dispatch.unsetBlockEditingMode( clientId )
+						);
+						sectionClientIdsInnerBlocks.forEach( ( clientId ) =>
+							dispatch.unsetBlockEditingMode( clientId )
+						);
+					}
```

Comment:
Can we add some e2e test to validate the expected behavior. The code here is a bit complex and I believe probably easy to break.

Reply:
Also, I think it's better to avoid calling setBlockEditingMode and unsetBlockEditingMode randomly in the "setMode" action. It's very adhoc. It can easily cause conflicts with other flows that enable/disable blocks (like template-locked mode...).

I believe "zoom-out" mode is something that the block editor package is aware of, so maybe we can instead bundle this behavior within the blocks (useBlockProps or wherever we consume the editing mode to disable stuff)

Reply:
I think @ellatrix had a similar suggestion: https://github.com/WordPress/gutenberg/pull/59249#discussion_r1542488671.

P.S. Previous comments get buried; it might be a good idea to squash commits even before PR is merged. cc @draganescu, @ajlende

Reply:
> Also, I think it's better to avoid calling setBlockEditingMode and unsetBlockEditingMode randomly in the "setMode" action. It's very adhoc. It can easily cause conflicts with other flows that enable/disable blocks (like template-locked mode...).

Yeah, I made a comment (that seems to have been buried) about keeping track of the initial state before switching so it can be restored. The series of `unsetBlockEditingMode` actions is pretty much equivalent to signaling the `'RESET_BLOCKS'` action as it is which doesn't account for changes that may happen in-between switching into `zoom-out` and switching out of `zoom-out`.

> I believe "zoom-out" mode is something that the block editor package is aware of, so maybe we can instead bundle this behavior within the blocks (useBlockProps or wherever we consume the editing mode to disable stuff)

I didn't like the idea of splitting up the logic that we already have for `'contentOnly'` in the store to another place. One of the few benefits we get from using a store for data is co-location of logic on that data.

Reply:
I moved this logic to the selector. I think it makes more sense there, although there are some more performance issues now that more logic has to be run on each select call. This code runs more that 10,000 times on a minimal page with a header, footer and a couple small groups in the content that has to be fixed now.

But at least it should avoid the problems with settings being overridden when you switch out of zoom-out mode.